Change compositing policy in response to memory pressure status

rreno opened the pull request:
#### 8299d7639dd907163642c19c331a29c98f34ba12
<pre>
Change compositing policy in response to memory pressure status
<a href="https://bugs.webkit.org/show_bug.cgi?id=259580">https://bugs.webkit.org/show_bug.cgi?id=259580</a>
rdar://109875865

Reviewed by Simon Fraser.

We switch to a more conservative compositing policy when the
MemoryPressureHandler has a change in memory usage policy at most every
2 seconds. In some cases we can enter memory warning and critical memory
pressure situations and get a foreground jetsam before this 2s has
expired. This change will respond to recent memory warning and memory
pressure notifications in order to change the compositing policy. While we
can&apos;t guarantee to not jetsam, this does improve our chances of
reducing layer counts before the system gets critically low on
memory.

* Source/WTF/wtf/MemoryPressureHandler.h:
(WTF::MemoryPressureHandler::isUnderMemoryWarning const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingPolicy):

Canonical link: <a href="https://commits.webkit.org/266497@main">https://commits.webkit.org/266497@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed9bff3c3a7bf04ebfda61f40dbd3bcae9875891

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14620 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15708 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13261 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14054 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14368 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15932 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14734 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16416 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12020 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12597 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19629 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11936 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12761 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15972 "3 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13241 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13308 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11171 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14015 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12582 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3633 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3386 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16914 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14401 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13150 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3445 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WTF/wtf/cocoa/MemoryPressureHandlerCocoa.mm:
```diff
@@ -61,7 +61,7 @@
 }
 
-static int notifyTokens[3];
+static int notifyTokens[6];
```
Could we provide some more details around where the number 6 comes from and what it represents.

stwrt commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -595,12 +595,12 @@ bool RenderLayerCompositor::updateCompositingPolicy()
+    if (cachedMemoryPressureStatus != isCurrentlyUnderPressure) {
         cachedMemoryPolicy = MemoryPressureHandler::singleton().currentMemoryUsagePolicy();
-        cachedMemoryPolicyTime = now;
+        cachedMemoryPressureStatus ^= true;
```
Why is this `^=`?

rreno replied:
This is a change of state for a two-state variable. I don’t know which way it’s going and xor-equals is a common way to do this bit flip.

stwrt commented on Source/WTF/wtf/MemoryPressureHandler.cpp:
```diff
@@ -268,6 +270,23 @@ void MemoryPressureHandler::endSimulatedMemoryPressure()
 
+void MemoryPressureHandler::beginSimulatedMemoryWarning()
+{
+    if (m_isSimulatingMemoryWarning)
```
It may be worth adding ASSERTS to track why we are calling `beginSimulatedMemoryWarning` after already enabling it. Same for `endSimulatedMemoryWarning`.

rreno replied:
I disagree! That would mean you could crash your process by sending the notification twice via `notifyutil`. Highly undesirable behavior.

rreno replied:
To answer the question you were implying (what is causing this to be called when we're already simulating) - if you post back-to-back notifications like this:

```
$ notifyutil -p org.WebKit.memoryWarning.begin
$ notifyutil -p org.WebKit.memoryWarning.begin
```
 (same with `.end` and with the `lowMemory` memory pressure sim variant)
 
 If we assert then you could easily crash your process by accidentally doing the above.

stwrt replied:
Ah, I didn't realize that the memory warning would stick around for a while and it could trigger it that easily.

stwrt reviewed (commented):
A few initial comments although I realize this is still in draft status.

Could we also have a test case to test this behavior.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -595,12 +595,12 @@ bool RenderLayerCompositor::updateCompositingPolicy()
-    static constexpr auto memoryPolicyCachingDuration = 2_s;
-    auto now = MonotonicTime::now();
-    if (now - cachedMemoryPolicyTime > memoryPolicyCachingDuration) {
+    static auto cachedMemoryPressureStatus = false;
```
I'm not sure initializing to false here, and the `^=` below is correct; what if it starts high? I think you should read `cachedMemoryPressureStatus` each time. Also, a bool called `status` is confusing; not sure what this represents, exactly.

rreno replied:
How about this?

```C++
static bool cachedIsUnderMemoryPressureOrWarning = MemoryPressureHandler::singleton().isUnderMemoryPressure() || MemoryPressureHandler::singleton().isUnderMemoryWarning();
```

Then I think the `^=` should be correct after that.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -593,14 +593,17 @@ bool RenderLayerCompositor::updateCompositingPolicy()
+    if (cachedIsUnderMemoryPressureOrWarning != isCurrentlyUnderMemoryPressureOrWarning()) {
         cachedMemoryPolicy = MemoryPressureHandler::singleton().currentMemoryUsagePolicy();
-        cachedMemoryPolicyTime = now;
+        cachedIsUnderMemoryPressureOrWarning ^= true;
```
I think you could avoid this confusing `^=` by writing this as:
```
bool nowUnderMemoryPressure = isCurrentlyUnderMemoryPressureOrWarning();
static bool isUnderMemoryPressureOrWarning = nowUnderMemoryPressure;

if (isUnderMemoryPressureOrWarning != nowUnderMemoryPressure) {
  cachedMemoryPolicy = MemoryPressureHandler::singleton().currentMemoryUsagePolicy();
  isUnderMemoryPressureOrWarning = nowUnderMemoryPressure;
}
```

rreno replied:
Cool, I like it. Though I don't think bit flips are all that confusing yours is definitely more readable.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -593,14 +593,18 @@ bool RenderLayerCompositor::updateCompositingPolicy()
-    static constexpr auto memoryPolicyCachingDuration = 2_s;
-    auto now = MonotonicTime::now();
-    if (now - cachedMemoryPolicyTime > memoryPolicyCachingDuration) {
+    auto nowUnderMemoryPressure = isCurrentlyUnderMemoryPressureOrWarning();
```
I don't like `auto` for `bool`.

rreno replied:
I'll push an amended commit before merge-queue -ing.

smfr reviewed (approved):
We should be on the lookout for ping-pongs between conservative compositing and not after this. Perhaps we should add some hysteresis going from conservative -> non-conservative to dampen the feedback loop?

rreno committed:
Change compositing policy in response to memory pressure status
https://bugs.webkit.org/show_bug.cgi?id=259580
rdar://109875865

Reviewed by Simon Fraser.

We switch to a more conservative compositing policy when the
MemoryPressureHandler has a change in memory usage policy at most every
2 seconds. In some cases we can enter memory warning and critical memory
pressure situations and get a foreground jetsam before this 2s has
expired. This change will respond to recent memory warning and memory
pressure notifications in order to change the compositing policy. While we
can't guarantee to not jetsam, this does improve our chances of
reducing layer counts before the system gets critically low on
memory.

* Source/WTF/wtf/MemoryPressureHandler.h:
(WTF::MemoryPressureHandler::isUnderMemoryWarning const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingPolicy):

Canonical link: https://commits.webkit.org/266497@main
